p2p/lightchain.py

```diff
@@ -245,35 +263,105 @@ def peer_block_height(peer: LESPeer) -> int:
         await self.wait_until_finished()
         self.logger.debug("LightPeerChain finished")
 
+    async def _wait_for_reply(self, request_id: int) -> Dict[str, Any]:
+        reply = None
+        got_reply = asyncio.Event()
+
+        def callback(r):
+            nonlocal reply
+            reply = r
+            got_reply.set()
+
+        self._pending_replies[request_id] = callback
+        await wait_with_token(got_reply.wait(), token=self.cancel_token, timeout=self.reply_timeout)
+        return reply
+
     @alru_cache(maxsize=1024, cache_exceptions=False)
-    async def get_block_header_by_hash(self, block_hash: Hash32) -> BaseBlock:
+    async def get_block_header_by_hash(self, block_hash: Hash32) -> BlockHeader:
         peer = await self.get_best_peer()
-        self.logger.debug("Fetching header %s from %s", encode_hex(block_hash), peer)
-        return await peer.get_block_header_by_hash(block_hash, self.cancel_token)
+        return await self._get_block_header_by_hash(peer, block_hash)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_block_body_by_hash(self, block_hash: Hash32) -> BaseBlock:
         peer = await self.get_best_peer()
         self.logger.debug("Fetching block %s from %s", encode_hex(block_hash), peer)
-        return await peer.get_block_by_hash(block_hash, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_block_bodies([block_hash], request_id)
```

Comment:
This pattern of generating the `request_id` independent of issuing the request is odd to me.  Is there a reason why `sub_proto.send_get_block_bodies` doesn't generate the `request_id` and return it?

Reply:
The only drawback I can think of is that these sub_proto methods would then no longer conform to the standard `send_*` method signature, which doesn't return anything. Filed #843 for it